[WIP] RelationshipMixin: Improve performance of Relationships parent

kbrock opened the pull request:
~~extracted #10299 - waiting to merge that to decide what would be the best to read here~~ MERGED

**Theme:**

Speed up `Capture.perf_capture_timer` (142s)
by speeding up `MiqAlert.target_needs_realtime_capture?` (105s)
by speeding up "detection if a resource has an alert" (105s)
by speeding up "get tags for a resource and it's parents (including blue folder)" (100s)
by speeding up "get parents for a resource" (~50% time, 80% queries)

**Goal of this PR:**

Speed up the time it takes to lookup a vm's parents by making it quicker to query ancestors/relationships. This speeds up `parent_blue_folder` and `parent_resource_pool`

https://bugzilla.redhat.com/show_bug.cgi?id=1346988
https://bugzilla.redhat.com/show_bug.cgi?id=1346999
## TL;DR

19% faster, 26% fewer objects, 17% fewer queries, 30% fewer rows returned.
## before

| ms | bytes | objects | queries | query (ms) | rows | `comments` |
| --: | --: | --: | --: | --: | --: | --- |
| 139,250.5 | 439,519,495* | 146,637,247 | 82,206 | 44,212.9 | 101,046 | `before` |
| 1,773.5 |  |  | 11 | 863.0 | 11,045 | `...Metric::Targets.capture_infra_targets` |
| 444.8 |  |  | 202 | 134.9 |  | `..Metric::Capture.calc_targets_by_rollup_parent` |
| 106.5 |  |  | 80 | 21.0 |  | `..Metric::Capture.calc_tasks_by_rollup_parent` |
| 101,671.1 |  |  | 60,223 | 29,461.1 | 90,000 | `..Metric::Capture.filter_perf_capture_now` |
| 7,350.9 |  |  | 5,221 | 5,908.7 | 10,000 | `...SELECT "taggings"."id" AS t0_r0, "taggings"."taggable_id" AS t0_r1, "taggings"."tag_id" AS t0_r2, "ta` |
| 4,896.8 |  |  | 10,000 | 3,701.4 | 25,000 | `...SELECT "relationships".*` |
| 10,203.7 |  |  | 30,000 | 6,794.4 | 30,000 | `...SELECT  "relationships".*` |
| 4,701.8 |  |  | 10,000 | 3,002.7 | 20,000 | `...SELECT "ems_folders".*` |
| 2,307.0 |  |  | 5,000 | 1,666.4 | 5,000 | `...SELECT "resource_pools".*` |
| 35,133.3 |  |  | 21,684 | 13,716.2 |  | `..Metric::Capture.queue_captures` |
- Memory usage does not reflect 71,630,466 freed objects. 
## after

| ms | bytes | objects | queries | query (ms) | rows | `comments` |
| --: | --: | --: | --: | --: | --: | --- |
| 112,703.9 | 245,611,833* | 108,651,445 | 57,206 | 36,877.4 | 71,046 | `after` |
| 1,828.7 |  |  | 11 | 877.2 | 11,045 | `...Metric::Targets.capture_infra_targets` |
| 441.7 |  |  | 202 | 132.4 |  | `..Metric::Capture.calc_targets_by_rollup_parent` |
| 112.1 |  |  | 80 | 24.1 |  | `..Metric::Capture.calc_tasks_by_rollup_parent` |
| 74,396.0 |  |  | 35,223 | 22,228.2 | 60,000 | `..Metric::Capture.filter_perf_capture_now` |
| 6,055.6 |  |  | 5,221 | 3,984.9 | 10,000 | `...SELECT "taggings"."id" AS t0_r0, "taggings"."taggable_id" AS t0_r1, "taggings"."tag_id" AS t0_r2, "ta` |
| 3,879.3 |  |  | 10,000 | 2,511.9 | 20,000 | `...SELECT "relationships"."ancestry"` |
| 6,282.9 |  |  | 10,000 | 5,148.0 | 10,000 | `...SELECT  "relationships".*` |
| 3,627.2 |  |  | 5,000 | 2,496.6 | 15,000 | `...SELECT "ems_folders".*` |
| 2,382.3 |  |  | 5,000 | 1,719.6 | 5,000 | `...SELECT  "resource_pools".*` |
| 35,785.4 |  |  | 21,684 | 13,598.4 |  | `..Metric::Capture.queue_captures` |
- Memory usage does not reflect 52,462,392 freed objects. 

I tried to cut out cruft from tables. let me know if you need more info

Fryguy commented on app/models/relationship.rb:
```diff
@@ -17,10 +17,16 @@ def filtered?(of_type, except_type)
+      relationships.reject { |r| r.filtered?(of_type, except_type) }
+    else
+      relationships = relationships.where(:resource_type => of_type) if of_type.present?
+      relationships = relationships.where.not(:resource_type => except_type) if except_type.present?
```
Minor, but since we have a separate method for `filtered?`, it feels like we should have a separate method for this scoping...perhaps a `.filtered` class method that does these two lines.

kbrock replied:
The purpose of this method `filter_by_resource_type` is to filter on these 2 fields.

You're proposing a method that just does the in memory part?

of note, `relationships.reject { |r| r.filtered?(of_type, except_type) }` is only used in this 1 spot.
Also, of note, `:except_type` is only ever used in 1 spot of our code.

Fryguy replied:
I'm proposing a method that is just the scopes part...  This method is essentially just a switch between scopes vs Array.

Fryguy replied:
Since it's a method that's just scopes, that in itself would make it a scope-method (i.e. a method that returns an AR::Relation)

kbrock replied:
moved, not sure why it isn't showing up in this commit. it is showing up correctly in main PR

Fryguy commented on spec/models/relationship_spec.rb:
```diff
@@ -23,4 +27,67 @@
+
+  def filter(relationships, options)
+    Relationship.filter_by_resource_type(relationships, options)
+  end
```
I'm not a fan of moving the method-under-test into a helper method...yes, it's less typing, but the separation is confusing.

kbrock replied:
fixed. not sure why it is showing up here

Fryguy commented:
Can you expand on the OP description a bit...those number are meaningless to me without context (e.g. is that 142s total time or 142s faster but what percentage is that)

Do you have some benchmarks for "Speed[ing] up the time it takes to lookup a vm's parents"?

Fryguy commented:
Overall code-wise looks good, but would like some benchmarks to justify it better.

Fryguy commented:
Thanks for the numbers @kbrock .  Sounds good.  Only thing left is my comment on separate method and getting the tests to go green

kbrock commented:
@Fryguy added scopes and cleared up dumb bug I introduced in fixing up cops

Fryguy commented:
@kbrock I see you made a lot more changes after I gave a :+1:,  and started moving more stuff around...the code was good as is, but now I have to re-review in light of all the new changes...were they necessary to move for this PR?

kbrock commented:
@Fryguy I'm sorry. The only changes that I think I made are related to your comments.

1. added a `filtered` scope per your suggestion
2. removed `filtered` helper method from specs per your suggestion
3. removed `assignment_mixin.rb` flat_map refactor that was unrelated, unnecessary, and was breaking the tests.

kbrock commented:
testing:

I ran for >10,000 Vms, with the old method and the new method. and they returned the same value. (not conclusive, but a little reassuring)
Though the Vms probably had the same parent lineage anyway

Fryguy commented on app/models/ems_folder.rb:
```diff
@@ -166,6 +166,23 @@ def folder_path_objs(*args)
   end
 
+  # similar to folder_path_objs, but takes in a scope
+  # @param folder [Relation] relation pointin to ems folder of interest
```
Typo `pointin` => `pointing`  Also prefer `EmsFolder` over `ems folder`

Fryguy commented on app/models/ems_folder.rb:
```diff
@@ -166,6 +166,23 @@ def folder_path_objs(*args)
+  def self.folder_path_objs(folder, *args)
+    options = args.extract_options!
+    folder_rels = Relationship.find_by(:resource_id => folder.select(:id), :resource_type => "EmsFolder")
+    if folder_rels.nil? # parent folder has no parents, use the parent folder
```
This doesn't seem right...or rather, the comment doesn't quite match the code in question.  A folder can have an entry in the Relationships table and also not have parents...this can happen if the folder is the root of a tree.  Granted, we don't have that case, but this code feels wrong from a general perspective.

However, there is another problem with this code and that is that it doesn't handle the relationship_type.  If the folder were to appear in multiple trees, this would not handle that case...or more specifically, it would randomly get only one of the tree entries because it's using find_by.

kbrock replied:
added support for `relationship_types` - good eye. We'll talk when we get back from vaca

Fryguy commented:
@kbrock If you remove the parent_blue_folder commit, then this would be fine to merge...that particular commit has issues that could probably be done in a separate PR.

I assume the numbers in the OP are without that commit?  If so, they seemed good to me without that parent_blue_folder commit.

Fryguy commented on app/models/mixins/relationship_mixin.rb:
```diff
@@ -109,6 +110,19 @@ def relationship_ids
+  # from this id, relationship records can be brought back and mapped to the resource of interest
+  # NOTE: parent_id is read from ancestry field, so it is not a db hit, but parent is an N+1 db hit.
+  def relationship_parent_ids
+    relationships.where.not(:ancestry => [nil, ""]).select(:ancestry).collect(&:parent_id)
```
I just realized...does this handle relationship_type filtering?

Fryguy commented on app/models/vm_or_template.rb:
```diff
@@ -791,21 +791,26 @@ def disconnect_storage(s = nil)
-    f = parent_blue_folder
-    f.nil? ? [] : f.folder_path_objs(*args)
+    with_relationship_type('ems_metadata') do
+      # parent_blue_folder # but as a scope
```
No need for this comment as it's redundant

Fryguy commented on app/models/mixins/relationship_mixin.rb:
```diff
@@ -109,6 +110,19 @@ def relationship_ids
+
+  # Returns relationships for the root relationships
+  # NOTE: can't cache this since it is a scope. caching relationship_parent_ids instead
+  def rel_roots(resource_type = {})
```
I don't quite understand what this is looking up...can you explain?  roots are just `.roots`, right?  or given a relationship, just call `.root`  What I don't understand is what roots have to do with parents.

kbrock commented:
status: extracged 10299, and probably discuss again the blue folder stuff with @Fryguy

kbrock commented:
majority of the meat has been extracted.
What remains is moving `parent_blue_folder` over to using scopes. This reduces 8 (12 rows) queries to 3 (6 rows) and shows most of the benefit

kbrock commented:
@Fryguy is there any legwork I can do on my side to make this easier to merge?
Or anything to extract to a separate PR?

thnx

Fryguy commented on app/models/mixins/relationship_mixin.rb:
```diff
@@ -135,6 +135,13 @@ def parent_ids(*args)
     Relationship.resource_pairs(parent_rels(*args))
   end
 
+  # return a scope of all of the parents of this record (must :of_type with a single parameter)
```
If of_type is required or single parameter, can you check and raise an ArgumentError (instead of the comment?)

Fryguy commented on app/models/vm_or_template.rb:
```diff
@@ -804,8 +804,9 @@ def parent_blue_folder
-    f = parent_blue_folder
-    f.nil? ? [] : f.folder_path_objs(*args)
+    with_relationship_type('ems_metadata') do
+      parents_scope(:of_type => "EmsFolder").folder_path_objs("ems_metadata", *args)
```
How does this scope get used? In the folder_path_objs, I don't see a reference to self anywhere, but maybe I'm missing something?

kbrock replied:
`EmsFolder` has:
```ruby
folder_path_resource_ids = folder_rels.path.where(:resource_type => "EmsFolder",
                                                  :relationship  => relationship_type).select(:resource_id)
```

The `folder_rels` is called on `parents_scope(:of_type => "EmsFolder")`

Fryguy commented on app/models/ems_folder.rb:
```diff
@@ -166,6 +166,26 @@ def folder_path_objs(*args)
+    end
+    # we can translate these into scopes if needed
+    folders = folders[1..-1] if options[:exclude_root_folder]
+    folders = folders.reject(&:hidden?) if options[:exclude_non_display_folders]
```
Since both of the methods have this part the same, can we move the common code into a private method?

Fryguy commented on app/models/ems_folder.rb:
```diff
@@ -160,10 +160,24 @@ def register_host(host)
-    folders = folders[1..-1] if options[:exclude_root_folder]
-    folders = folders.reject(&:hidden?) if options[:exclude_non_display_folders]
-    folders
+    self.class.folder_options path(:of_type => "EmsFolder"), options
```
I'm surprised the bot is not picking up on the lack of parens on the method call (i.e. please use parens).

Fryguy commented on app/models/ems_folder.rb:
```diff
@@ -160,10 +160,24 @@ def register_host(host)
+                                                           .select(:resource_id)
+                EmsFolder.where(:id => folder_path_resource_ids)
+              end
+    folder_options folders, options
```
Same

Fryguy commented on app/models/ems_folder.rb:
```diff
@@ -195,4 +209,11 @@ def self.child_folder_paths_recursive(subtree, options = {})
   end
   private_class_method :child_folder_paths_recursive
+
+  def self.folder_options(folders, options)
```
Can you name this "apply_folder_options".  "folder_options" is a noun, so implies you are asking for some object.  "apply_" turns it into a verb making it clear than an action is being taken on the params.  You *may* want to make it "apply_folder_options!" to signify that it's modifying the params.

kbrock replied:
don't think it is modifying the `options` or `folders`. like calling it `apply_folder_options`

Fryguy commented on app/models/mixins/relationship_mixin.rb:
```diff
@@ -134,6 +134,16 @@ def parent_ids(*args)
   end
 
+  # return a scope of all of the parents of this record
+  def parents_scope(*args)
```
If we want parents to be a scope, can we just changes parents to do this directly?

Fryguy commented on app/models/mixins/relationship_mixin.rb:
```diff
@@ -134,6 +134,16 @@ def parent_ids(*args)
+    options = args.extract_options!
+    if options[:of_type].nil? || options[:of_type].kind_of?(Array)
+      raise ArgumentError, ":of_type required as a single parameter"
+    end
```
Not sure I like forcing this down the of_type path.  I know that's your current use case, but from a pure API perspective I would never expect that.

Fryguy commented:
@kbrock Would like to sit down with you to understand this one like we did last time.  I liked that last time we came to nice elegant answer that covered all use cases, and I have a feeling the same thing will happen here.

kbrock commented:
WIP until I rebase and produce numbers again
